### PR TITLE
Team/seed export

### LINKED ISSUE #4233: Extension - Seed extraction

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #4233

Needs luniehq/cosmos-keys#74

**Description:**

Work In Progress

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by faboweb** [comment]
🚀 Deployed on https://5ee3deb0f21c88b8aed011c0--lunieio.netlify.app

**Comment by Bitcoinera** [comment]
My idea with the "Edit" and "Delete" button is to display the a Modal with the "Feature not yet available message". What do you think?

**Comment by Bitcoinera** [comment]
It is working but not sure how to get the seed for the previously stored wallets...

**Comment by Bitcoinera** [comment]
TODO:

- RevealSeedModal should be actually a child component of AccountModal, since buttons and input will be commont to RevealSeedModal and DeleteAccountModal. Same idea that in ActionModal with SendModal, DelegationModal, and so on.

- Create a new folder in `components` called `account`. Move all these new modals there. Add the necessary paths (like in `jest.config.js`)

- Improve close functions. Display seed. Improve "Feature not available" display. Improve the switch logic between `Delete` & `Show` & `Edit`

**Comment by faboweb** [comment]
Expected behavior:
You click on the dots on the right and the account list item moves to the left revealing the buttons. There is no mentioning of balances. Let's leave that out for another time.

**Comment by faboweb** [comment]
We can't reveal seeds in the extension. We don't store the seeds and we can't create the seed from the private key.
We can show to users the privateKey if there is no seedPhrase present. What do you think?

**Comment by faboweb** [comment]
several improvements:
- use route for revealing and not a modal to keep the logic small
- fixed account list logic
- fixed wallet extraction
- added wallet displaying

still needed:
reveal password and seed view page need to be styled according to sketch designs

**Comment by Bitcoinera** [comment]
> We can't reveal seeds in the extension. We don't store the seeds and we can't create the seed from the private key.
> We can show to users the privateKey if there is no seedPhrase present. What do you think?

ah, nice idea! How is it possible that I didn't realize! But then we need to also allow recover from privKey in Lunie right.

Should be very straightforward though

**Comment by Bitcoinera** [comment]
> * use route for revealing and not a modal to keep the logic small

Actually, I realized later that day that I have made a mistake and that in the designs you can see that the "reveal your seed" thing is not another modal, but the same modal than AccountList transforming. Should have added that to the first TODO

**Comment by Bitcoinera** [comment]
This is working now!

image

Amazing, I didn't know at first what you meant with the 'transition to the left when you click', but now I saw it. I love this solution!

**Comment by Bitcoinera** [comment]
This is how RevealSeedModal is looking now:

**Comment by Bitcoinera** [comment]
Getting this for all the accounts?

image

**Comment by Bitcoinera** [comment]
Thought maybe being possible to import via private key in Polkadot it would work too for Cosmos, but seems not yet:

File: app/src/components/common/TmSessionImportName.vue [hunk 1]
```diff
@@ -101,17 +101,30 @@ export default {
     },
   },
   async created() {
-    this.importedAddress = await this.$store.dispatch(`getAddressFromSeed`, {
-      seedPhrase: this.$store.state.recover.seed,
-      network: this.network,
-    })
+    if (this.isSeedOrPrivateKey(this.$store.state.recover.seed)) {
+      this.importedAddress = await this.$store.dispatch(
+        `recoverFromPrivatekey`,
```
Comment (faboweb) [inline app/src/components/common/TmSessionImportName.vue:106]: out of scope

**Comment by faboweb** [comment]
Implementation is not matching designs.
Import private key is not part of this issue.

File: app/src/components/common/TmSessionImportName.vue (hunk 1 above)
Comment (Bitcoinera) [reply app/src/components/common/TmSessionImportName.vue:106]: :crying_cat_face:

**Comment by Bitcoinera** [comment]
> Import private key is not part of this issue.

OK, I will remove that part

> Implementation is not matching designs.

Which parts exactly? The sliding to the left was something that we just decided to change no? Please be more specific

File: app/src/vuex/modules/keystore.js [hunk 2]
```diff
@@ -74,6 +74,19 @@ export default () => {
...
+      const privateKeyBuffer = Buffer.from(privateKey, `hex`)
+
+      // get current network
+      const networkObject = await getNetworkInfo(network, store)
+      const recoveredAddress = derivePublicFromPrivateKey(
+        privateKeyBuffer,
+        networkObject.bech32_prefix
+      )
+      return recoveredAddress
+    },
```
Comment (Bitcoinera) [inline app/src/vuex/modules/keystore.js:89]: OK, but this one I can leave it right :sweat_smile: For the next PR :v:

**Comment by faboweb** [comment]
> Implementation is not matching designs.

I was talking about the views for entering the password and showing the seed
https://www.sketch.com/s/8ffba04e-5055-44dd-984c-c60a74f5103b/a/lyDJ87

Spacing needs to be adjusted. The coloring of the words would be nice. It would be great if the buttons would look like this, but not if the consistancy of the UI would suffer.

**Comment by Bitcoinera** [comment]
> I was talking about the views for entering the password and showing the seed
> https://www.sketch.com/s/8ffba04e-5055-44dd-984c-c60a74f5103b/a/lyDJ87

Oops, true, too much empty space down there:

**Comment by Bitcoinera** [comment]
What about now?

**Comment by Bitcoinera** [comment]
But yeah, somehow extension is not being detected by FE anymore, debuggin 🤔

**Comment by faboweb** [comment]
@mariopino can you take over the styling please

File: app/changes/team_seed-export [hunk 3]
```diff
@@ -0,0 +1 @@
+[Added] [#4233](https://github.com/cosmos/lunie/issues/4233) Seed export @mariopino
```
Comment (Bitcoinera) [inline app/changes/team_seed-export:1]: another usurper! xD it should say team 😉

Comment (mariopino) [reply app/changes/team_seed-export:1]: Ops, sorry!!! I forgot!

**Comment by faboweb** [comment]
@mariopino can you share screenshot? hard to preview this

**Comment by mariopino** [comment]
> @mariopino can you share screenshot? hard to preview this

Sure!

image

image

image

image

image

image

**Comment by faboweb** [comment]
Amazing work @mariopino !!!

**Comment by Bitcoinera** [comment]
Good to go! 💃

**Comment by Bitcoinera** [comment]
Ah, the snaps

**Comment by Bitcoinera** [comment]
I love this PR so much! How we worked all together on it and how wonderful it turned out 🤩 

Team power!!! 💪 🔥

File: extension/src/store/actions.js [hunk 4]
```diff
@@ -3,6 +3,7 @@ import gql from 'graphql-tag'
 import { NetworksAll } from '../popup/gql'
 import { lunieMessageTypes, parseTx } from '../scripts/parsers'
 import { storeWallet } from '@lunie/cosmos-keys'
+// import { resolve } from 'core-js/fn/promise'
```
Comment (faboweb) [inline extension/src/store/actions.js:6]:
```suggestion
```

File: app/src/vuex/modules/keystore.js [hunk 5]
```diff
@@ -65,6 +74,19 @@ export default () => {
...
```
Comment (faboweb) [inline app/src/vuex/modules/keystore.js:89]:
```suggestion
```

File: app/src/components/account/AccountMenu.vue [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          class="account-menu-button account-menu-show-seed"
+          :to="{ name: 'reveal', params: { address } }"
+        >
+          <i class="material-icons notranslate">visibility</i>
+        </router-link>
+        <span class="account-menu-button-span">Show Seed</span>
+      </div>
+    </div>
+    <!-- <RevealSeedModal v-if="command === `show-seed`" :address="address" />
+    <AccountModal v-else-if="command !== '' && command !== `show-seed`" /> -->
```
Comment (faboweb) [inline app/src/components/account/AccountMenu.vue:15]:
```suggestion
```

**Comment by Bitcoinera** [comment]
Let's merge!! :woman_dancing: 

I will open a new issue for implementing recovery via private key

**Pull request merged**